GUIDELINES.md

```diff
@@ -0,0 +1,135 @@
+* Create a test class for the _feature under test_ (e.g. `LoginTest`), so the test is agnostic of the underlying structures (classes, models, views, data structures, ...)
+  * Should it be that a test really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelTest`)
+* Create a steps class for the feature under test (e.g. `LoginSteps`) which will contain all the `given`, `when` and `then` functions as well as dependency configuration and setup code
+  * The steps class should know about the setup of the test, so for example if the steps class tailored to test the integration of multiple classes, it should also be called something like `LoginIntegrationSteps` which tests a broad stack of classes (e.g. a `LoginViewModel` talking to an `AuthManager` and a `SessionStore`)
+  * Should it be that a steps class really just tests _one_ class, feel free to name it exactly like that (`LoginViewModelSteps`)
+  * In case the steps class' aim is purely to offer and act on a fake implementation of a class, name it with a `FakeSteps` suffix (e.g. `BackendGatewayFakeSteps` or `SessionStoreFakeSteps`)
+* Start with the test
+  * In the test class you created add reference to the steps class with the `steps` function (e.g. `val sut by steps<LoginSteps>`) and call the value `sut` (system under test)
+  * Fill up the test with `@Test`-annotated test functions
+    * The test function can be named after a _scenario_ (e.g. `New user trying to log in`, `Weather condition is cloudy`)
+    * Only if necessary to distinguish test cases, also add more context like what's roughly expected (e.g. `New user trying to login - welcome to the app message shown`, `Existing user trying to login - welcome back message shown`)
+    * Don't use names like `givenNewUser_whenLoggingIn_thenShowWelcomeToAppMessage` because
+      * The test code in the test functions should already be easy enough to understand preconditions, actions and assumptions
+      * The more text you use the harder to distinguish test cases
+      * The less spaces you use the harder to read
+      * So please just use the test function names to make a meaningful distinctions to properly navigate between and understand test scenarios and cases
+  * For each test case, call functions that would reside in the steps class and prefix these functions with `given`, `when` or `then`
+    * Functions starting with `given` should set up the environment, as these define the assumptions that are made (e.g. `givenUser(email = "test@test.com", password = "supersecure")` indicating that the fake backend should act as if there was that user saved in the database). For the naming choose present tense (e.g. `givenUser` as a short form of "there is a user" or `givenInternetConnected`)
```

Comment:
We talked about this already. Can you point me to some resources, which describe the given-when-then like this? I don't think it should be differentiated by tense. Given=pre-conditions / When=action which is test / Then = expected state after when. That's how understand this pattern, see https://archive.is/20140814232159/http://www.jroller.com/perryn/entry/given_when_then_and_how#selection-335.1-176.81

Reply:
there are a lot of different opinions on that, e.g. in [fowler's article](https://martinfowler.com/bliki/GivenWhenThen.html) you linked once the "i" form is used. this is just a proposal, i'm for streamlining this. on jroller.com also present tense is used for "given" statements and actually i see nothing noteworthy that conflicts with his proposal. what do you prefer?

Reply:
I think given should describe pre-conditions of the test (which can be actions too). When = actions of the scenario. And then = the expected state afterwards. So, when we consider given defines the pre-conditions for the test, it can very well be also in the past tense. E.g. givenNeverLoggedIn could be also a precondition. 

I just don't think enforcing a tense should be in the guidelines. We can of course start to do it, but the more important part is that these keywords are used in the proper context of the test. An example would help.

Reply:
i think also this is a candidate for _not_ having it in the guidelines. in the end the guidelines should also be quick to understand, so it helps to not overcomplicate things.